File: src/main/cpp/g0001_0100/s0008_string_to_integer_atoi/Solution.cpp
// #Medium #Top_Interview_Questions #String #2024_05_23_Time_0_ms_(100.00%)_Space_8.2_MB_(36.95%)

#include <string>
#include <climits>

class Solution {
public:
    int myAtoi(std::string str) {
        if (str.empty()) {
            return 0;
        }
        int i = 0;
        bool negativeSign = false;
        int len = str.length();
        
        while (i < len && str[i] == ' ') {
            i++;
        }
        
        if (i == len) {
            return 0;
        }
        
        if (str[i] == '+') {
            i++;
        } else if (str[i] == '-') {
            i++;
            negativeSign = true;
        }
        
        int num = 0;
        while (i < len && str[i] >= '0' && str[i] <= '9') {
            int tem = str[i] - '0';
            if (negativeSign) {
                tem = -tem;
            }
            
            if (num < INT_MIN / 10 || (num == INT_MIN / 10 && tem < -8)) {
                return INT_MIN;
            } else if (num > INT_MAX / 10 || (num == INT_MAX / 10 && tem > 7)) {
                return INT_MAX;
            } else {
                num = num * 10 + tem;
                i++;
            }
        }
        return num;
    }
};

File: src/main/swift/g0001_0100/s0008_string_to_integer_atoi/Solution.swift
// #Medium #Top_Interview_Questions #String #2024_06_18_Time_0_ms_(100.00%)_Space_16.5_MB_(45.57%)

class Solution {
    func myAtoi(_ str: String) -> Int {
        if str.isEmpty {
            return 0
        }
        
        var i = str.startIndex
        var negativeSign = false
        var num = 0
        
        // Skip leading whitespace
        while i < str.endIndex && str[i] == " " {
            i = str.index(after: i)
        }
        
        if i == str.endIndex {
            return 0
        }
        
        // Check for sign
        if str[i] == "+" {
            i = str.index(after: i)
        } else if str[i] == "-" {
            i = str.index(after: i)
            negativeSign = true
        }
        
        while i < str.endIndex, let digit = str[i].wholeNumberValue {
            let tem = negativeSign ? -digit : digit
            
            // Check for overflow
            if num > Int32.max / 10 || (num == Int32.max / 10 && tem > 7) {
                return Int(Int32.max)
            }
            if num < Int32.min / 10 || (num == Int32.min / 10 && tem < -8) {
                return Int(Int32.min)
            }
            
            num = num * 10 + tem
            i = str.index(after: i)
        }
        
        return num
    }
}
